panic when receiving MsgAppResp from removed member

We check whether the message comes from removed member in etcdserver. But when it goes into node loop/raft, it may have been removed from the cluster. We need to check whether it exists in node loop.

```
?       github.com/coreos/etcd/etcdserver/etcdserverpb  [no test files]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x2f7804]

goroutine 1702 [running]:
runtime.panic(0x527240, 0x852e04)
    /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/coreos/etcd/raft.stepLeader(0xc208043760, 0x4, 0xcadce0ad0cf6289f, 0xe4656d48eba82c23, 0x2, 0x0, 0x10, 0x0, 0x0, 0x0, ...)
    /Users/yichengq/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:505 +0x4d4
github.com/coreos/etcd/raft.(*raft).Step(0xc208043760, 0x4, 0xcadce0ad0cf6289f, 0xe4656d48eba82c23, 0x2, 0x0, 0x10, 0x0, 0x0, 0x0, ...)
    /Users/yichengq/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:432 +0x141
github.com/coreos/etcd/raft.(*node).run(0xc2084a7130, 0xc208043760)
    /Users/yichengq/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:256 +0x59e
created by github.com/coreos/etcd/raft.StartNode
    /Users/yichengq/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:154 +0x5f7
```

--- a/raft/node.go
+++ b/raft/node.go
@@ -253,7 +253,10 @@ func (n *node) run(r *raft) {
 			m.From = r.id
 			r.Step(m)
 		case m := <-n.recvc:
-			r.Step(m) // raft never returns an error
+			// filter out response message from unknow From.
+			if _, ok := r.prs[m.From]; ok || !IsResponseMsg(m) {
+				r.Step(m) // raft never returns an error
+			}
 		case cc := <-n.confc:
 			if cc.NodeID == None {
 				r.resetPendingConf()
@@ -322,9 +325,7 @@ func (n *node) Tick() {
 	}
 }
 
-func (n *node) Campaign(ctx context.Context) error {
-	return n.step(ctx, pb.Message{Type: pb.MsgHup})
-}
+func (n *node) Campaign(ctx context.Context) error { return n.step(ctx, pb.Message{Type: pb.MsgHup}) }
 
 func (n *node) Propose(ctx context.Context, data []byte) error {
 	return n.step(ctx, pb.Message{Type: pb.MsgProp, Entries: []pb.Entry{{Data: data}}})
@@ -332,7 +333,7 @@ func (n *node) Propose(ctx context.Context, data []byte) error {
 
 func (n *node) Step(ctx context.Context, m pb.Message) error {
 	// ignore unexpected local messages receiving over network
-	if m.Type == pb.MsgHup || m.Type == pb.MsgBeat {
+	if IsLocalMsg(m) {
 		// TODO: return an error?
 		return nil
 	}
@@ -365,9 +366,7 @@ func (n *node) step(ctx context.Context, m pb.Message) error {
 	}
 }
 
-func (n *node) Ready() <-chan Ready {
-	return n.readyc
-}
+func (n *node) Ready() <-chan Ready { return n.readyc }
 
 func (n *node) Advance() {
 	select {

--- a/raft/util.go
+++ b/raft/util.go
@@ -44,6 +44,10 @@ func max(a, b uint64) uint64 {
 	return b
 }
 
+func IsLocalMsg(m pb.Message) bool { return m.Type == pb.MsgHup || m.Type == pb.MsgBeat }
+
+func IsResponseMsg(m pb.Message) bool { return m.Type == pb.MsgAppResp || m.Type == pb.MsgVoteResp }
+
 // DescribeMessage returns a concise human-readable description of a
 // Message for debugging.
 func DescribeMessage(m pb.Message) string {

